capstone/capweb/templatetags/cached_json_object.py

```diff
@@ -0,0 +1,80 @@
+import json
+
+from django.utils.safestring import mark_safe
+from django import template
+from django.db import connections
+from django.core.cache import cache
+from django.conf import settings
+
+from capdb.models import Reporter, Jurisdiction, Court
+
+register = template.Library()
+
+@register.simple_tag()
+def cached_json_object(label, default="", force_clear=False):
+    """
+        Return contents of named cached_json_object, from cache if possible. If cached_json_object is not found, return default.
+    """
+    cached_json_object_router = {
+        'map_numbers': map_numbers,
+        'search_jurisdiction_list': search_jurisdiction_list,
+        'search_court_list': search_court_list,
+        'search_reporter_list': search_reporter_list,
+        'court_abbrev_list': court_abbrev_list,
+    }
+    key = 'cached_json_object:{}'.format(label)
+    cached_json_object = cache.get(key)
+
+    if force_clear or (not cached_json_object and label in cached_json_object_router):
+        cache.set(key, cached_json_object_router[label](), settings.CACHED_JSON_OBJECT_TIMEOUT)
+        cached_json_object = cache.get(key)
+
+    return mark_safe(cached_json_object if cached_json_object else default)
+
+
+def map_numbers():
+    cursor = connections['capdb'].cursor()
+    cursor.execute(r"""
+        SELECT 
+          j.slug,
+          COUNT(c.id) AS case_count,
+          COUNT(DISTINCT c.volume_id) AS volume_count,
+          COUNT(DISTINCT c.reporter_id) AS reporter_count,
+          SUM(CASE WHEN (c.first_page||c.last_page)~E'^\\d+$' THEN c.last_page::integer-c.first_page::integer+1 ELSE 1 END) AS page_count
+        FROM capdb_jurisdiction j 
+          LEFT JOIN capdb_casemetadata c ON j.id=c.jurisdiction_id 
+        WHERE
+          c.in_scope IS True
+        GROUP BY j.id;
+    """)
```

Comment:
This takes 100 seconds to run, so it doesn't work to do it in a view. I think this is like the tsv files -- we want to add it to a `/download/statistics/latest/jurisdiction_counts.json` file and load it from there.

Reply:
> This takes 100 seconds to run, so it doesn't work to do it in a view. I think this is like the tsv files -- we want to add it to a `/download/statistics/latest/jurisdiction_counts.json` file and load it from there.

Right— Ah, ok. I need to get shell access squared away so I can profile some of this stuff. My initial plan to mitigate this problem— which I didn't end up implementing— was having a function in our prod deployment routine that would hit any page using a particularly long-running query, which should populate Redis, right? I don't think we'd need to automatically expire these things because they shouldn't need to be updated any more frequently than snippets. If Redis ended up losing its state for some reason, it would at least be a self-fixing problem even if not immediately (though we should probably log instances of cached values being regenerated so we could monitor it.) Does that make sense or should I just scrap it?

Reply:
Yeah I think that's going to be jankier since it has a bunch of assumptions and moving parts built in -- assumes that redis isn't cleared frequently which is otherwise harmless, assumes that the caching step will complete even when the downstream request is killed after 1 minute, implies at least a couple minutes of error messages during regeneration (which in turn have to be checked to make sure they're not real errors), etc. Writing the file out to /download/statistics/ avoids any jank, and also comes with the upside that the file is potentially useful to people.

Reply:
> Writing the file out to /download/statistics/ avoids any jank, and also comes with the upside that the file is potentially useful to people.

Easy enough!